### PR TITLE
Split massa-consensus

### PR DESCRIPTION
**Pull request by adrien-zinger**
Split the consensus into multiple crates
worker, exports and also massa-graph and proof-of-stake-exports

This PR contains also the development of a split of configs and errors between POS, graph, ledger, and consensus.

I didn't manage the split of the existing unit tests to simplify the review. I would prefer to move the unit tests from massa-consensus-worker with another PR. This choice has an impact on the visibility of some structures and functions, and on some _test feature_ flags.

### CODE REVIEW

**Comment by adrien-zinger** [comment]
Currently, I've an issue with the tempfiles, I'm investigating

**Review by yvan-sraka — DISMISSED** [review]

File: massa-graph/src/active_block.rs
Comment (gterzian) [inline massa-graph/src/active_block.rs:50]: Good example of something we can address with the refactoring, to avoid cloning...

**Review by gterzian — APPROVED** [review]
Great, thanks! Agree on doing tests as follow-up...

File: massa-consensus-exports/src/tools.rs
Comment (AureliaDolo) [inline massa-consensus-exports/src/tools.rs:1]: These are tools meant to be only used in tests, I'd rename/move the file to make that clear

File: massa-graph/src/settings.rs
Comment (AureliaDolo) [inline massa-graph/src/settings.rs:17]: So you duplicate thread count in all the configs that need it right ?

File: massa-consensus-exports/src/settings.rs
Comment (AureliaDolo) [inline massa-consensus-exports/src/settings.rs:414]: Probable a discussion for another issue : shouldn't we get rid of the serialization context ?

File: massa-consensus-exports/src/settings.rs
Comment (AureliaDolo) [inline massa-consensus-exports/src/settings.rs:433]: why this import is not as the top of the file ?

**Review by AureliaDolo — COMMENTED** [review]
Why are pos.md and endorsement.md in the worker ? (mAybe in more of a hacking or internal section of the wiki)

**Comment by yvan-sraka** [comment]
> Why are pos.md and endorsement.md in the worker ? (maybe in more of a hacking or internal section of the wiki)

It's both something we like to have included in `cargo doc` output and linked in a hacking/internal section of wiki :)

**Review by damip — APPROVED** [review]
Didn't read all the details, but the general scheme seems good !

**Comment by yvan-sraka** [comment]
bors merge

File: massa-consensus-exports/src/tools.rs
Comment (gterzian) [reply massa-consensus-exports/src/tools.rs:1]: I think that's a good point, best to move those to the worker where the tests are.

**Comment by gterzian** [comment]
Since there are still unresolved discussions and suggested changes, we should wait for these to be resolved before trying to merge.

Also as a general note, after having addressed a review, it's best to ask for a second review, as opposed to assuming that the changes made were as the reviewer intended and to immediately try to merge. It's possible that the reviewer still has a comment, or that the changes do not completely address the earlier review.

cc @massalabs/core-team

**Comment by AureliaDolo** [comment]
> Since there are still unresolved discussions and suggested changes, we should wait for these to be resolved before trying to merge.
> 
> Also as a general note, after having addressed a review, in most cases it's a good idea to ask for a second review, as opposed to assuming that the changes made were as the reviewer intended and to immediately try to merge. It's possible that the reviewer still has a comment after the changes, for example if they do not completely address the earlier review comments.
> 
> The exception would be for really minor or even optional changes, which the reviewer can make clear by marking the comments with "nit", or perhaps by using the ["LGTM with comments"](https://google.github.io/eng-practices/review/reviewer/speed.html#lgtm-with-comments) convention.
> 
> We could consider eventually writing down clearly in our own doc what standards we want to apply to our own review process, I've opened an issue for this at #2220.
> 
> cc @massalabs/core-team

Related to #1069 and #1927

File: massa-graph/src/settings.rs
Comment (adrien-zinger) [reply massa-graph/src/settings.rs:17]: The thread count management will be the object of another issue/PR

File: massa-graph/src/active_block.rs
Comment (adrien-zinger) [reply massa-graph/src/active_block.rs:50]: Can you clarify your thoughts? :) We could see if we can move the ActiveBlock in models or use a kind of code as @yvan-sraka showed up in the meet today:

```rust
struct Grounded;
struct Launched;

struct Rocket<T = Grounded> {
    stage: std::marker::PhantomData<T>,
}

impl Rocket<Grounded> {
    pub fn launch(self) -> Rocket<Launched> {}
}

impl Rocket<Launched> {
    pub fn accelerate(&mut self) {}
    pub fn decelerate(&mut self) {}
}

impl<T> Rocket<T> {
    pub fn color(&self) -> Color {}
    pub fn weight(&self) -> Kilograms {}
}
```

File: massa-graph/src/settings.rs
Comment (AureliaDolo) [reply massa-graph/src/settings.rs:17]: #2197

File: massa-consensus-exports/src/settings.rs
Comment (AureliaDolo) [reply massa-consensus-exports/src/settings.rs:414]: Probably linked to #2197 too

**Review by AureliaDolo — DISMISSED** [review]

File: massa-consensus-exports/src/tools.rs
Comment (adrien-zinger) [reply massa-consensus-exports/src/tools.rs:1]: Actually, one of these functions is also used in an implementation of ConsensusConfig (the `impl<&Path>`) that cannot be moved easily. We should first think about _"Where we should put the common configs tools and implementations"_. I'm open to any proposals!!

File: massa-graph/src/active_block.rs
Comment (adrien-zinger) [reply massa-graph/src/active_block.rs]: Ok, I just pushed one version with the atoms of the code in models, ⁣:wink:  what do you think?

**Review by gterzian — DISMISSED** [review]
LGTM with one comment

**Comment by AureliaDolo** [comment]
It looks like you also tackled #237 here ?

**Review by yvan-sraka — DISMISSED** [review]
LGTM

File: massa-graph/src/active_block.rs
Comment (adrien-zinger) [reply massa-graph/src/active_block.rs]: @gterzian Do you think you can close this discussion? :)

File: massa-consensus-exports/src/settings.rs
Comment (adrien-zinger) [reply massa-consensus-exports/src/settings.rs:433]: I fixed that point forward with the config refactoring

File: massa-graph/src/active_block.rs
Comment (adrien-zinger) [reply massa-graph/src/active_block.rs]: I resolve it because the code has changed. Nevertheless, we'll keep the discussion open in future PR.

File: massa-models/src/error.rs
Comment (adrien-zinger) [reply massa-models/src/error.rs:39]: I close that and open the #2257

**Review by AureliaDolo — APPROVED** [review]

**Comment by adrien-zinger** [comment]
bors merge

**Pull request merged**
